daemon/api_apps_test.go

```diff
@@ -771,3 +778,104 @@ func (s *appsSuite) TestLogsNoServices(c *check.C) {
 	rspe := s.errorReq(c, req, nil)
 	c.Assert(rspe.Status, check.Equals, 404)
 }
+
+func (s *appsSuite) TestLogsFromNamespaces(c *check.C) {
+	s.expectLogsAccess()
+
+	st := s.d.Overlord().State()
+	st.Lock()
+	err := servicestatetest.MockQuotaInState(st, "foo", "", []string{"snap-a"}, quota.NewResourcesBuilder().WithJournalSize(16*quantity.SizeMiB).Build())
+	st.Unlock()
+	c.Assert(err, check.IsNil)
+
+	s.jctlRCs = []io.ReadCloser{
+		ioutil.NopCloser(strings.NewReader(`
+{"MESSAGE": "hello1", "SYSLOG_IDENTIFIER": "xyzzy", "_PID": "42", "__REALTIME_TIMESTAMP": "42"}
+{"MESSAGE": "hello2", "SYSLOG_IDENTIFIER": "xyzzy", "_PID": "42", "__REALTIME_TIMESTAMP": "44"}
+{"MESSAGE": "hello3", "SYSLOG_IDENTIFIER": "xyzzy", "_PID": "42", "__REALTIME_TIMESTAMP": "46"}
+{"MESSAGE": "hello4", "SYSLOG_IDENTIFIER": "xyzzy", "_PID": "42", "__REALTIME_TIMESTAMP": "48"}
+{"MESSAGE": "hello5", "SYSLOG_IDENTIFIER": "xyzzy", "_PID": "42", "__REALTIME_TIMESTAMP": "50"}
+`))}
+
+	req, err := http.NewRequest("GET", "/v2/logs?names=snap-a.svc2&n=42&follow=false", nil)
+	c.Assert(err, check.IsNil)
+
+	rec := httptest.NewRecorder()
+	s.req(c, req, nil).ServeHTTP(rec, req)
+
+	c.Check(s.jctlSvcses, check.DeepEquals, [][]string{{"snap.snap-a.svc2.service"}})
+	c.Check(s.jctlNmspcs, check.DeepEquals, []string{"snap-foo"})
+	c.Check(s.jctlNs, check.DeepEquals, []int{42})
+	c.Check(s.jctlFollows, check.DeepEquals, []bool{false})
+
+	c.Check(rec.Code, check.Equals, 200)
+	c.Check(rec.Header().Get("Content-Type"), check.Equals, "application/json-seq")
+	c.Check(rec.Body.String(), check.Equals, `
+{"timestamp":"1970-01-01T00:00:00.000042Z","message":"hello1","sid":"xyzzy","pid":"42"}
+{"timestamp":"1970-01-01T00:00:00.000044Z","message":"hello2","sid":"xyzzy","pid":"42"}
+{"timestamp":"1970-01-01T00:00:00.000046Z","message":"hello3","sid":"xyzzy","pid":"42"}
+{"timestamp":"1970-01-01T00:00:00.000048Z","message":"hello4","sid":"xyzzy","pid":"42"}
+{"timestamp":"1970-01-01T00:00:00.00005Z","message":"hello5","sid":"xyzzy","pid":"42"}
+`[1:])
+}
+
+func (s *appsSuite) TestLogsFromMultipleNamespaces(c *check.C) {
```

Comment:
Do we have/need a test that checks if the "follow" mode of the journal code also works?

Reply:
Yes definitely - but I think I need to revise this approach a bit, it looks like this approach might have broken follow mode entirely. Because we now do intermittent reading and close the logs, follow mode no longer can work.